### DISCUSSION TITLE
Data Storage for LuaLink

### DISCUSSION THREAD

**Discussion by tut13** [LuaLink/LuaLinkV1#30] [General]
LuaLink is a great plugin, especially if you are a lua main(like me), but functions/commands are quite missing IMO, this is not a feature request but rather a chat, has any of you made script functions/commands for example in Denizen you can [flag](https://meta.denizenscript.com/Docs/Search/flag#flag)an object with data. And I would like to ask for ways you have done this with LuaLink

**Comment by Saturn745** [comment]
With LuaLink, you can create [commands](https://github.com/LuaLink/LuaLink/blob/main/docs/home.md#commands) and [event listeners](https://github.com/LuaLink/LuaLink/blob/main/docs/home.md#events). Nearly anything achievable with a typical Bukkit plugin can be accomplished through a Lua script. Regarding "flagging an object with data," I'm not entirely clear, as I haven't used Denizen before. Correct me if I'm mistaken, but it appears provide some sort of data storage. If that's correct, you can utilize [PDC](https://docs.papermc.io/paper/dev/pdc). A simple example of PDC can be found in the [grappling hook](https://github.com/LuaLink/Scripts/blob/main/Fun/grappling.lua) script I created, which uses PDC to set values for the grappling hook and its cooldown. The [message](https://github.com/LuaLink/Scripts/blob/main/Utility/message.lua) script also uses PDC to indicate whether a player can receive messages. If this isn't the case, could you please provide further clarification?

**Reply by tut13** [reply]
*i'm terrible at giving descriptions*

 I was talking about "methods" like the [ban method](https://hub.spigotmc.org/javadocs/bukkit/org/bukkit/entity/Player.html#ban(java.lang.String,java.time.Duration,java.lang.String,boolean)) that you can use on players, I'm used to Denizen which is basically a readable scripting language and their commands work a little differently and are very specific to Denizen but can easily be achieved in LuaLink, I just wasn't smart enough to think about it that way. 

 The flag command is basically the same as PDC, thank you so much for the reply! Helped a lot!

**Reply by Saturn745** [reply]
Just about every method available in Purpur is available since that is the API LuaLink uses. So for example:
```lua
script.registerSimpleCommand(function(sender, args)
    if utils.instanceOf(sender, "org.bukkit.entity.Player") then
        sender:ban("You have been banned for using the /example command", nil, nil, true)
    end
end, {
        name = "example",
})
```
This just registers a command `/example` when ran it checks if the sender is a player, if they are it bans them.

**Reply by tut13** [reply]
Yeah, I didn't really have much experience directly with Purpur API, and was quite confused. Thank you for the explanation!
